File: ios/KeyAddressManager.m
//
//  KeyAddressManager.m
//  buoyantwallet
//
//  Created by Kristofer Pelchat on 8/14/19.
//  Copyright © 2019 Facebook. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "KeyAddressManager.h"
#import <React/RCTLog.h>
#import <Keyaddr/Keyaddr.h>

@implementation KeyAddressManager

RCT_EXPORT_MODULE();

RCT_REMAP_METHOD(keyaddrWordsFromBytes, lang:(NSString*)lang bytes:(NSString*)bytes
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  
  NSString *words = KeyaddrWordsFromBytes(lang, bytes, error);
  if (error) {
    reject(@"no_events", @"Issue calling keyaddrWordsFromBytes", *error);
  } else {
    resolve(words);
  }
}

RCT_REMAP_METHOD(keyaddrWordsToBytes, lang:(NSString*)lang words:(NSString*)words
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  
  NSString *bytes = KeyaddrWordsToBytes(lang, words, error);
  if (error) {
    reject(@"no_events", @"Issue calling keyaddrWordsToBytes", *error);
  } else {
    resolve(bytes);
  }
}

RCT_REMAP_METHOD(keyaddrWordsFromPrefix, lang:(NSString*)lang prefix:(NSString*)prefix max:(nonnull NSNumber*)max
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  
  NSString *words = KeyaddrWordsFromPrefix(lang, prefix, [max longValue]);
  if (error) {
    reject(@"no_events", @"Issue calling keyaddrWordsFromPrefix", *error);
  } else {
    resolve(words);
  }
}

RCT_REMAP_METHOD(newKey, seed:(NSString*)seed
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  KeyaddrKey *keyAddrKey = KeyaddrNewKey(seed, error);
  
  if (error) {
    reject(@"no_events", @"Issue calling newKey", *error);
  } else {
    resolve([keyAddrKey key]);
  }
}

RCT_REMAP_METHOD(child, key:(NSString*)key index:(nonnull NSNumber*)index
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  KeyaddrKey *keyAddrKey = KeyaddrFromString(key, error);
  KeyaddrKey *privateKey = [keyAddrKey child:[index intValue] error:error];
  
  if (error) {
    reject(@"no_events", @"Issue calling child", *error);
  } else {
    resolve([privateKey key]);
  }
}

RCT_REMAP_METHOD(hardenedChild,keyHC:(NSString*)keyHC index:(nonnull NSNumber*)index
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  KeyaddrKey *keyAddrKey = KeyaddrFromString(keyHC, error);
  KeyaddrKey *privateKey = [keyAddrKey hardenedChild:[index intValue] error:error];
  
  if (error) {
    reject(@"no_events", @"Issue calling hardenedChild", *error);
  } else {
    resolve([privateKey key]);
  }
}

RCT_REMAP_METHOD(toPublic,key:(NSString*)key
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  KeyaddrKey *keyAddrKey = KeyaddrFromString(key, error);
  KeyaddrKey *publicKey = [keyAddrKey toPublic:error];
  
  if (error) {
    reject(@"no_events", @"Issue calling toPublic", *error);
  } else {
    resolve([publicKey key]);
  }
}

RCT_REMAP_METHOD(ndauAddress,ndauAddr:(NSString*)key
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  KeyaddrKey *keyAddrKey = KeyaddrFromString(key, error);
  KeyaddrAddress *address = [keyAddrKey ndauAddress:error];
  
  if (error) {
    reject(@"no_events", @"Issue calling ndauAddress", *error);
  } else {
    resolve([address address]);
  }
}

RCT_REMAP_METHOD(sign,key:(NSString*)key msgstr:(NSString*)msgstr
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  KeyaddrKey *keyAddrKey = KeyaddrFromString(key, error);
  KeyaddrSignature *sig = [keyAddrKey sign:msgstr error:error];
  RCTLogInfo(@"sign call created signature %@", sig);
  
  if (error) {
    reject(@"no_events", @"Issue calling sign", *error);
  } else {
    resolve([sig signature]);
  }
}

RCT_REMAP_METHOD(deriveFrom,parentKey:(NSString*)parentKey parentPath:(NSString*)parentPath childPath:(NSString*)childPath
                 findEventsWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject)
{
  NSError *__autoreleasing *error = NULL;
  KeyaddrKey *keyAddrKey = KeyaddrDeriveFrom(parentKey, parentPath, childPath, error);
  
  if (error) {
    reject(@"no_events", @"Issue calling deriveFrom", *error);
  } else {
    resolve([keyAddrKey key]);
  }
}

@end

File: android/app/src/main/java/io/oneiro/buoyantwallet/KeyAddressManager.java
package io.oneiro.buoyantwallet;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import keyaddr.Key;
import keyaddr.Keyaddr;

public class KeyAddressManager extends ReactContextBaseJavaModule {
    public KeyAddressManager(ReactApplicationContext reactApplicationContext) {
        super(reactApplicationContext);
    }

    @Override
    public String getName() {
        return "KeyAddressManager";
    }

    @ReactMethod
    public void keyaddrWordsFromBytes(String lang, String bytes, Promise promise) {
        try {
            promise.resolve(Keyaddr.wordsFromBytes(lang, bytes));
        } catch (Exception e) {
            promise.reject("problem getting words from bytes", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void keyaddrWordsToBytes(String lang, String words, Promise promise) {
        try {
            promise.resolve(Keyaddr.wordsToBytes(lang, words));
        } catch (Exception e) {
            promise.reject("problem getting words to bytes", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void keyaddrWordsFromPrefix(String lang, String prefix, Integer max, Promise promise) {
        try {
            promise.resolve(Keyaddr.wordsFromPrefix(lang, prefix, max));
        } catch (Exception e) {
            promise.reject("problem getting words from prefix", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void newKey(String bytes, Promise promise) {
        try {
            promise.resolve(Keyaddr.newKey(bytes).getKey());
        } catch (Exception e) {
            promise.reject("problem getting newKey", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void child(String key, Integer childIndex, Promise promise) {
        try {
            Key theKey = Keyaddr.fromString(key);

            promise.resolve(theKey.child(childIndex).getKey());
        } catch (Exception e) {
            promise.reject("problem getting child", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void hardenedChild(String key, Integer childIndex, Promise promise) {
        try {
            Key theKey = Keyaddr.fromString(key);

            promise.resolve(theKey.hardenedChild(childIndex).getKey());
        } catch (Exception e) {
            promise.reject("problem getting child", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void toPublic(String key, Promise promise) {
        try {
            Key theKey = Keyaddr.fromString(key);

            promise.resolve(theKey.toPublic().getKey());
        } catch (Exception e) {
            promise.reject("problem getting neuter", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void ndauAddress(String key, Promise promise) {
        try {
            Key theKey = Keyaddr.fromString(key);

            promise.resolve(theKey.ndauAddress().getAddress());
        } catch (Exception e) {
            promise.reject("problem getting ndauAddress", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void sign(String key, String message, Promise promise) {
        try {
            Key theKey = Keyaddr.fromString(key);

            promise.resolve(theKey.sign(message).getSignature());
        } catch (Exception e) {
            promise.reject("problem getting ndauAddress", e.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void deriveFrom(String parentKey, String parentPath, String childPath, Promise promise) {
        try {
            Key theKey = Keyaddr.deriveFrom(parentKey, parentPath, childPath);

            promise.resolve(theKey.getKey());
        } catch (Exception e) {
            promise.reject("problem getting deriveFrom", e.getLocalizedMessage());
        }
    }
}
